src/DefinitionResolver.php

```diff
@@ -670,23 +670,23 @@ public function resolveExpressionNodeToType($expr)
                 if ($expr->parent instanceof Node\Expression\CallExpression) {
                     $add .= '()';
                 }
-                $fqn = $classFqn . $add;
-                $def = $this->index->getDefinition($fqn);
-                if ($def !== null) {
-                    return $def->type;
-                } else {
-                    $classDef = $this->index->getDefinition($classFqn);
-                    if ($classDef !== null && is_array($classDef->extends)) {
-                        foreach ($classDef->extends as $parent) {
-                            $def = $this->index->getDefinition($parent . $add);
-                            if ($def !== null) {
-                                if ($def->type instanceof Types\This) {
-                                    return new Types\Object_(new Fqsen('\\' . $classFqn));
-                                }
-                                return $def->type;
-                            }
+                $lookupDefinition = function (string $base, string $add) use (&$lookupDefinition) {
```

Comment:
Yes, it should be trivial to move the generator to a method. I just wanted to get an OK on the approach first.